### PR TITLE
[#12] [Backend] Build a core networking service

### LINKED ISSUE #12: [Backend] Build a core networking service
## Why

To benefit the developers for the networking implementation, we should build a core network layer.

## Acceptance criteria

Building a core network layer that fulfills these technical goals:

- provide a simple interface to make network requests; attach basic headers to network requests; also provide customization points
- provide basics of response data decoding
- provide basics of a response error decoding
- provide an API that makes it easy to stub/mock network requests for testing purposes

Integrate the core network layer to support the following build environments:

- Staging: https://nimble-survey-web-staging.herokuapp.com/
- Production: https://survey-api.nimblehq.co/

## Resourses

- [Nimble survey API](https://documenter.getpostman.com/view/11835486/T1LJjTqj)
- The application follows the [standard JSON API spec v1](https://jsonapi.org/)

### PR DESCRIPTION
**Pull request by thiennguyen0196**
- Close #12

## What happened 👀

- Add necessary libraries into build gradle, including Koin, Ktor, Napier, Json API, Settings, Timber, Preference, Coroutines
- Create Api HTTP client, service, repository, response & request model for refreshing token API for example

## Insight 📝

N/A

## Proof Of Work 📹

The app can still be run with Android devices

### CODE REVIEW

**Review by nmint8m — APPROVED** [review]
@thiennguyen0196 Seems that this PR is also for #13? Please also link this PR to that issue.

**Comment by thiennguyen0196** [comment]
@nmint8m this has not been tested for #13 yet, when doing the #13 I will test it carefully again and link it if it can run and qualify the requrements

File: shared/src/commonMain/kotlin/co/nimblehq/kaylabruce/kmmic/data/local/LocalService.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    val accessToken: String?
+
+    val refreshToken: String?
+
+    val expiresIn: Int?
+    fun saveToken(token: Token)
+    fun clear()
+}
+
+class LocalServiceImpl(private val settings: Settings) : LocalService {
```
Comment (markgravity) [inline shared/src/commonMain/kotlin/co/nimblehq/kaylabruce/kmmic/data/local/LocalService.kt:25]: What do you think about `LocalSettingsService`? The current name is kindly generic 🙏

File: shared/src/commonMain/kotlin/co/nimblehq/kaylabruce/kmmic/data/remote/service/AuthService.kt [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import co.nimblehq.kaylabruce.kmmic.data.remote.model.request.RefreshTokenRequest
+import co.nimblehq.kaylabruce.kmmic.data.remote.model.response.TokenResponse
+
+interface AuthService {
+    suspend fun refreshToken(body: RefreshTokenRequest): TokenResponse
+}
+
+class AuthServiceImpl(private val apiClient: ApiClient) : AuthService {
+
+    override suspend fun refreshToken(body: RefreshTokenRequest): TokenResponse {
```
Comment (markgravity) [inline shared/src/commonMain/kotlin/co/nimblehq/kaylabruce/kmmic/data/remote/service/AuthService.kt:12]: Just curious, why do we implement this first? it's not `auth/token`?

Comment (thiennguyen0196) [reply shared/src/commonMain/kotlin/co/nimblehq/kaylabruce/kmmic/data/remote/service/AuthService.kt:12]: the `auth/token` will be implemented in this [ticket](https://github.com/nimblehq/ic-kmm-kayla-bruce/issues/13) @markgravity Currently I just implement a random function to have the example calling of API service 😄

File: shared/src/commonMain/kotlin/co/nimblehq/kaylabruce/kmmic/data/local/LocalService.kt (hunk 1 above)
Comment (thiennguyen0196) [reply shared/src/commonMain/kotlin/co/nimblehq/kaylabruce/kmmic/data/local/LocalService.kt:25]: To me the name of service should be generic, as right now we are using `Settings` to implement it but what if we decided to change the dependency 🤔 then we have to change the name of service class too which is not recommended 
But if we keep the name of `LocalService` interface and just change the name of implementation class to `LocalSettingServiceImpl` then I am okay with it 👌

Comment (markgravity) [reply shared/src/commonMain/kotlin/co/nimblehq/kaylabruce/kmmic/data/local/LocalService.kt]: But the `LocalService` name here is quite confusing, right? We don't know what it is used for. 🤔 
Something like `LocalStorageService` makes more sense IMO

Comment (thiennguyen0196) [reply shared/src/commonMain/kotlin/co/nimblehq/kaylabruce/kmmic/data/local/LocalService.kt]: I agreed, let me change it to `LocalStorageService` 👍

**Comment by nmint8m** [comment]
@markgravity Is there any concern from your side?

**Review by markgravity — APPROVED** [review]

**Pull request merged**